### PR TITLE
[win/asan] Search both higher and lower in AllocateTrampolineRegion

### PR DESCRIPTION
**Pull request by zmodem**
There may not always be available virtual memory at higher addresses than the target function. Therefore, search also lower addresses while ensuring that we stay within the accessible memory range.

Additionally, add more ReportError calls to make the reasons for interception failure more clear.

### CODE REVIEW

File: compiler-rt/lib/interception/interception_win.cpp [hunk 1]
```diff
@@ -340,32 +347,66 @@ struct TrampolineMemoryRegion {
...
+UNUSED static const uptr kTrampolineRangeLimit = 1ull << 31;  // 2 gig
 static const int kMaxTrampolineRegion = 1024;
 static TrampolineMemoryRegion TrampolineRegions[kMaxTrampolineRegion];
 
-static void *AllocateTrampolineRegion(uptr image_address, size_t granularity) {
-#if SANITIZER_WINDOWS64
-  uptr address = image_address;
-  uptr scanned = 0;
-  while (scanned < kTrampolineScanLimitRange) {
+static void *AllocateTrampolineRegion(uptr min_addr, uptr max_addr,
```
Comment (rnk) [inline compiler-rt/lib/interception/interception_win.cpp:354]: I think this deserves a significant doc comment. The loop below probes the address space in a search that alternates between forward probing and backwards probing, and that makes the logic fairly hard to follow. I'd stop short of ASCII art, but a text list like this would help:
```
// This loop probes the virtual address space to find free memory in range to implement a trampoline. It alternates searching backwards and forwards, probing regions below in the order P1, P2, P3, P4, etc.
// min_addr
// ...
// P4
// P2
// func_addr
// P1
// P3
// ...
// max_addr
```

File: compiler-rt/lib/interception/interception_win.cpp [hunk 2]
```diff
@@ -340,32 +347,66 @@ struct TrampolineMemoryRegion {
...
+                                      uptr func_addr, size_t granularity) {
+#  if SANITIZER_WINDOWS64
+  // Clamp {min,max}_addr to the accessible address space.
+  SYSTEM_INFO system_info;
+  ::GetSystemInfo(&system_info);
```
Comment (rnk) [inline compiler-rt/lib/interception/interception_win.cpp:359]: The call site here calls `GetMmapGranularity`, which also calls `GetSystemInfo`. Can we do `GetSystemInfo` once outside the loop, and store this information about the address space (mmap_granularity, addrspace_start, addrspace_end) in global variables? The performance gain is marginal, it's mostly to have fewer parameters and locals in this function.

File: compiler-rt/lib/interception/interception_win.cpp [hunk 3]
```diff
@@ -340,32 +347,66 @@ struct TrampolineMemoryRegion {
...
 
-    // Check whether a region can be allocated at |address|.
+    // Check whether a region can be allocated at |addr|.
     if (info.State == MEM_FREE && info.RegionSize >= granularity) {
-      void *page = ::VirtualAlloc((void*)RoundUpTo(address, granularity),
-                                  granularity,
-                                  MEM_RESERVE | MEM_COMMIT,
-                                  PAGE_EXECUTE_READWRITE);
+      void *page =
+          ::VirtualAlloc((void *)addr, granularity, MEM_RESERVE | MEM_COMMIT,
```
Comment (rnk) [inline compiler-rt/lib/interception/interception_win.cpp:393]: Super nit: There is a race here between VirtualAlloc and VirtualQuery. Nothing prevents other threads (or processes) from tampering with the virtual address space during our search process. If we want this loop to be bulletproof, we would want some kind of retry case here. I think that's probably unnecessary, but this seems like a good place to insert a `ReportError` call to provide more context on failed VirtualAlloc calls.

File: compiler-rt/lib/interception/interception_win.cpp (hunk 1 above)
Comment (zmodem) [reply compiler-rt/lib/interception/interception_win.cpp:354]: Added a comment that should make it more clear.

It doesn't go as far as providing a concrete example like you suggested, because the order would depend on the addresses, and at that point I think the example would be more complex than it's worth.

File: compiler-rt/lib/interception/interception_win.cpp (hunk 2 above)
Comment (zmodem) [reply compiler-rt/lib/interception/interception_win.cpp:359]: Agreed that it shouldn't matter for performance. `GetSystemInfo` should be fast, and the `AllocateTrampolineRegion` call (along with the preceding `GetMmapGranularity`) only happens 1-3 times per process startup. (While this is inside a loop, this "if" branch is taken at most once per loop.)

I'm not sure global variables would be simpler.

Both `AllocateTrampolineRegion` and its caller (`AllocateMemoryForTrampoline`) need the mmap granularity, so it has to be passed from one to the other somehow. I suppose the caller could pass a reference to a SYSTEM_INFO object instead, but I'm not sure that's simpler really.

I'd like to leave this as-is for now, and would be happy to follow up if folks have ideas for simplification.

File: compiler-rt/lib/interception/interception_win.cpp (hunk 3 above)
Comment (zmodem) [reply compiler-rt/lib/interception/interception_win.cpp:400]: Thanks, I didn't think about that. Added a ReportError call.

**Comment by zmodem** [comment]
I've updated the patch. PTAL.

**Review by rnk — APPROVED** [review]
Looks good, thanks!
